File: collector/src/toolCallHistory.test.ts
import { describe, it, expect } from 'vitest';
import { createEmptyHistory, updateHistory, toProjectRelative } from './toolCallHistory.js';
import type { TranscriptEvent } from './transcriptParser.js';

function assistantEvent(toolUseId: string, toolName: string, filePath: string | null, timestamp: Date, sessionId: string | null = null): TranscriptEvent {
  return {
    kind: 'assistant', sessionId, timestamp, cwd: null, model: null, usage: null,
    toolUses: [{ id: toolUseId, name: toolName, input: filePath ? { file_path: filePath } : {} }],
    toolResults: [], humanText: null, originKind: null,
  };
}

function userResultEvent(toolUseId: string, timestamp: Date, sessionId: string | null = null): TranscriptEvent {
  return {
    kind: 'user', sessionId, timestamp, cwd: null, model: null, usage: null,
    toolUses: [], toolResults: [{ toolUseId, resultLength: 10 }], humanText: null, originKind: null,
  };
}

describe('collector toolCallHistory', () => {
  it('opens a tool call on tool_use and closes it on the matching tool_result', () => {
    const t0 = new Date('2026-07-28T00:00:00Z');
    const t1 = new Date('2026-07-28T00:00:01Z');
    let history = createEmptyHistory();
    history = updateHistory(history, [assistantEvent('tu_1', 'Read', 'src/foo.ts', t0)], t0.getTime());
    expect(history.events).toEqual([]);
    expect(history.openByToolUseId['tu_1']).toEqual({ toolName: 'Read', filePath: 'src/foo.ts', startedAt: t0.getTime(), subagentType: null, sessionId: null });

    history = updateHistory(history, [userResultEvent('tu_1', t1)], t1.getTime());
    expect(history.events).toEqual([
      { toolUseId: 'tu_1', toolName: 'Read', filePath: 'src/foo.ts', startedAt: t0.getTime(), closedAt: t1.getTime() },
    ]);
    expect(history.openByToolUseId['tu_1']).toBeUndefined();
  });

  it('captures subagent_type from Agent tool_use input', () => {
    const t0 = new Date('2026-07-28T00:00:00Z');
    let history = createEmptyHistory();
    history = updateHistory(history, [{
      kind: 'assistant', sessionId: 'sess_123', timestamp: t0, cwd: null, model: null, usage: null,
      toolUses: [{ id: 'tu_agent', name: 'Agent', input: { subagent_type: 'general-purpose' } }],
      toolResults: [], humanText: null, originKind: null,
    }], t0.getTime());
    expect(history.openByToolUseId['tu_agent']).toEqual({
      toolName: 'Agent',
      filePath: null,
      startedAt: t0.getTime(),
      subagentType: 'general-purpose',
      sessionId: 'sess_123',
    });
  });

  it('leaves subagentType null when subagent_type is missing from input', () => {
    const t0 = new Date('2026-07-28T00:00:00Z');
    let history = createEmptyHistory();
    history = updateHistory(history, [{
      kind: 'assistant', sessionId: 'sess_456', timestamp: t0, cwd: null, model: null, usage: null,
      toolUses: [{ id: 'tu_read', name: 'Read', input: { file_path: 'src/foo.ts' } }],
      toolResults: [], humanText: null, originKind: null,
    }], t0.getTime());
    expect(history.openByToolUseId['tu_read']).toEqual({
      toolName: 'Read',
      filePath: 'src/foo.ts',
      startedAt: t0.getTime(),
      subagentType: null,
      sessionId: 'sess_456',
    });
  });

  it('captures sessionId for every tool_use regardless of tool name', () => {
    const t0 = new Date('2026-07-28T00:00:00Z');
    let history = createEmptyHistory();
    history = updateHistory(history, [{
      kind: 'assistant', sessionId: 'sess_789', timestamp: t0, cwd: null, model: null, usage: null,
      toolUses: [
        { id: 'tu_agent', name: 'Agent', input: { subagent_type: 'special-agent' } },
        { id: 'tu_edit', name: 'Edit', input: { file_path: 'test.ts' } },
      ],
      toolResults: [], humanText: null, originKind: null,
    }], t0.getTime());
    expect(history.openByToolUseId['tu_agent'].sessionId).toBe('sess_789');
    expect(history.openByToolUseId['tu_edit'].sessionId).toBe('sess_789');
  });

  it.runIf(process.platform === 'win32')(
    'rejects a cross-drive absolute path that relative() cannot make relative (Windows only)',
    () => {
      // On win32, path.relative() between paths on different drives returns
      // the unchanged absolute `to` path with no '..' segments, so the
      // traversal check alone would let this slip through un-rejected.
      const result = toProjectRelative(String.raw`D:\Secrets\Matt\x.ts`, String.raw`C:\Users\Matt\projects\foo`);
      expect(result).toBeNull();
    },
  );
});

File: collector-go/internal/transcript/toolcallhistory_test.go
package transcript

import (
	"runtime"
	"testing"
	"time"
)

func assistantToolUseEvent(toolUseID, toolName string, filePath *string, timestamp time.Time) Event {
	input := map[string]interface{}{}
	if filePath != nil {
		input["file_path"] = *filePath
	}
	return Event{
		Kind:      "assistant",
		Timestamp: &timestamp,
		ToolUses:  []ToolUse{{ID: toolUseID, Name: toolName, Input: input}},
	}
}

func userResultEvent(toolUseID string, timestamp time.Time) Event {
	return Event{
		Kind:        "user",
		Timestamp:   &timestamp,
		ToolResults: []ToolResult{{ToolUseID: toolUseID, ResultLength: 10}},
	}
}

func TestUpdateHistory_OpensAndClosesOnMatchingToolResult(t *testing.T) {
	t0 := time.Date(2026, 7, 28, 0, 0, 0, 0, time.UTC)
	t1 := time.Date(2026, 7, 28, 0, 0, 1, 0, time.UTC)

	history := CreateEmptyHistory()
	history = UpdateHistory(history, []Event{assistantToolUseEvent("tu_1", "Read", strp("src/foo.ts"), t0)}, t0.UnixMilli())
	if len(history.Events) != 0 {
		t.Fatalf("expected no closed events yet, got %d", len(history.Events))
	}
	open, ok := history.OpenByToolUseID["tu_1"]
	if !ok {
		t.Fatalf("expected tu_1 to be open")
	}
	if open.ToolName != "Read" || open.FilePath == nil || *open.FilePath != "src/foo.ts" || open.StartedAt != t0.UnixMilli() {
		t.Fatalf("unexpected open entry: %+v", open)
	}

	history = UpdateHistory(history, []Event{userResultEvent("tu_1", t1)}, t1.UnixMilli())
	if len(history.Events) != 1 {
		t.Fatalf("expected 1 closed event, got %d", len(history.Events))
	}
	closed := history.Events[0]
	if closed.ToolUseID != "tu_1" || closed.ToolName != "Read" || closed.FilePath == nil || *closed.FilePath != "src/foo.ts" {
		t.Fatalf("unexpected closed event: %+v", closed)
	}
	if closed.StartedAt != t0.UnixMilli() || closed.ClosedAt != t1.UnixMilli() {
		t.Fatalf("unexpected timestamps: %+v", closed)
	}
	if _, ok := history.OpenByToolUseID["tu_1"]; ok {
		t.Fatalf("expected tu_1 to no longer be open")
	}
}

func TestToProjectRelative_CrossDriveWindowsOnly(t *testing.T) {
	if runtime.GOOS != "windows" {
		t.Skip("windows-only cross-drive guard")
	}
	// On win32, path.relative()/filepath.Rel() between paths on different
	// drives cannot produce a usable relative path -- must be rejected, not
	// silently passed through as an absolute path.
	result := ToProjectRelative(strp(`D:\Secrets\Matt\x.ts`), strp(`C:\Users\Matt\projects\foo`))
	if result != nil {
		t.Fatalf("expected nil for cross-drive path, got %q", *result)
	}
}

func TestToProjectRelative_PosixStyleAbsoluteAgainstWindowsRoot(t *testing.T) {
	if runtime.GOOS != "windows" {
		t.Skip("windows-only: exercises filepath.IsAbs vs Node's win32 isAbsolute divergence")
	}
	// A POSIX-style path (no drive letter) is absolute per Node's win32
	// path.isAbsolute (leading separator alone suffices), so it must be
	// routed through Rel-and-traversal-guard and rejected -- never passed
	// through unredacted just because Go's filepath.IsAbs says it's relative.
	result := ToProjectRelative(strp(`/home/matt/secret.ts`), strp(`C:\projects\x`))
	if result != nil {
		t.Fatalf("expected nil for POSIX-style absolute path against Windows root, got %q", *result)
	}
}

func TestToProjectRelative_CaseInsensitiveDriveLetter(t *testing.T) {
	if runtime.GOOS != "windows" {
		t.Skip("windows-only: exercises drive-letter case normalization")
	}
	// Node's win32 path.relative is case-insensitive on the drive letter;
	// filepath.Rel is not. A lowercase-drive file path against an
	// uppercase-drive root should still relativize successfully.
	result := ToProjectRelative(strp(`c:\projects\x\src\a.ts`), strp(`C:\projects\x`))
	if result == nil {
		t.Fatalf("expected non-nil result for case-varied drive letter")
	}
	if *result != `src\a.ts` {
		t.Fatalf("expected %q, got %q", `src\a.ts`, *result)
	}
}

func TestToProjectRelative_RelativizesUnderRoot(t *testing.T) {
	root := `/home/matt/projects/foo`
	abs := `/home/matt/projects/foo/src/bar.ts`
	if runtime.GOOS == "windows" {
		root = `C:\Users\Matt\projects\foo`
		abs = `C:\Users\Matt\projects\foo\src\bar.ts`
	}
	result := ToProjectRelative(&abs, &root)
	if result == nil {
		t.Fatalf("expected non-nil result")
	}
}

func TestToProjectRelative_RejectsTraversal(t *testing.T) {
	p := "../../secret"
	result := ToProjectRelative(&p, strp("/whatever"))
	if result != nil {
		t.Fatalf("expected nil for traversal path, got %q", *result)
	}
}

func TestToProjectRelative_NilFilePath(t *testing.T) {
	if ToProjectRelative(nil, strp("/root")) != nil {
		t.Fatalf("expected nil")
	}
}

func TestToProjectRelative_AbsoluteWithNoRoot(t *testing.T) {
	p := "/some/abs/path"
	if runtime.GOOS == "windows" {
		p = `C:\some\abs\path`
	}
	if ToProjectRelative(&p, nil) != nil {
		t.Fatalf("expected nil when projectRoot is nil")
	}
	empty := ""
	if ToProjectRelative(&p, &empty) != nil {
		t.Fatalf("expected nil when projectRoot is empty")
	}
}

func TestUpdateHistory_CapsAtHistoryMaxEvents(t *testing.T) {
	history := CreateEmptyHistory()
	base := time.Date(2026, 7, 28, 0, 0, 0, 0, time.UTC)
	for i := 0; i < HistoryMaxEvents+10; i++ {
		id := "tu_" + string(rune('a'+i%26)) + string(rune(i))
		ts := base.Add(time.Duration(i) * time.Second)
		history = UpdateHistory(history, []Event{assistantToolUseEvent(id, "Read", nil, ts)}, ts.UnixMilli())
		history = UpdateHistory(history, []Event{userResultEvent(id, ts)}, ts.UnixMilli())
	}
	if len(history.Events) != HistoryMaxEvents {
		t.Fatalf("expected history capped at %d, got %d", HistoryMaxEvents, len(history.Events))
	}
}
